### PR TITLE
no more paranoid db models

### PR DESCRIPTION
**Pull request by iamigo**
Remove "paranoid" setting from db models (exc. Users)
Remove isDeleted attribute (exc. Users)
Add migration to clean up the soft-deleted data and update the indices
Update tests

Note: we are keeping paranoid on for Users because we can't delete some of the soft-deleted user records due to referential integrity, where another record has a soft-deleted user associated with it as creator and/or owner.

Changes to /api/v1/helpers/nouns/*:
- pull "isDeleted" out of the list of "readOnlyFields"

Changes to /db/model/*:
- remove "paranoid" setting on model
- remove "isDeleted" field defn
- no longer call setIsDeleted in beforeDestroy
- remove "isDeleted" from list of fields in any index defns
- remove "isDeleted" from list of "ignoreAttributes" when calling publishObject

Migration:

Up
1. fixReferences - clear the collectorGroupId col for collectors which reference a soft-deleted collector group
1.  destroySoftDeleted - deletes any soft-deleted records from the db
1. removeOldIndices - deletes the indices which include deletedAt/isDeleted
1. removeFields - removes the deletedAt/isDeleted columns
1. dropAndAddUniqueIndices - drops and adds new indices without deletedAt/isDeleted - we have to do this here because the db tries to create the new unique indices on db startup, but fails because the soft-deleted records are still there, so now that they're gone, we can drop and recreate these

Down
1. recreateOldFields - adds back the deletedAt/isDeleted columns
1. removeNewIndices - removes the new indices without deletedAt/isDeleted, if they were created

### CODE REVIEW

File: api/v1/controllers/ssoconfig.js [hunk 1]
```diff
@@ -80,22 +80,23 @@ module.exports = {
...
-        const err = new apiErrors.ResourceNotFoundError();
-        err.info = 'There is no sso config to delete.';
-        u.handleError(next, err, helper.modelName);
+        found = o;
+        return o.destroy();
       }
+
+      const err = new apiErrors.ResourceNotFoundError();
+      err.info = 'There is no sso config to delete.';
+      u.handleError(next, err, helper.modelName);
```
Comment (jgraff2) [inline api/v1/controllers/ssoconfig.js]: Won't this still allow the next "then" block to execute?

File: db/model/user.js [hunk 2]
```diff
@@ -198,12 +195,7 @@ module.exports = function user(seq, dataTypes) {
             seq.models.Profile.findById(_this.profileId)
             .then((p) => {
               if (p) {
-                if (Number(p.isDeleted) === 0) {
-                  resolve();
-                } else {
-                  reject(new Error(
-                    `Profile ${_this.profileId} has been soft-deleted`));
-                }
+                resolve();
```
Comment (jgraff2) [inline db/model/user.js]: ok to change this even though we're keeping paranoid for users?

File: tests/db/model/aspect/delete.js [hunk 3]
```diff
@@ -49,13 +45,10 @@ describe('tests/db/model/aspect/delete.js', () => {
...
       .then((o) => o.destroy())
-      .then((o) => {
-        if (o.deletedAt && o.isDeleted) {
-          done();
-        } else {
-          done(new Error('expecting it to be soft-deleted'));
-        }
-      })
+      .then((o) => Aspect.findOne({ where: { name: u.name } }))
+      .then((o) => Aspect.findOne({ where: { name: u.name } }))
```
Comment (jgraff2) [inline tests/db/model/aspect/delete.js]: duplicate line

File: migrations/20190520221710-no-paranoid.js [hunk 4]
```diff
@@ -0,0 +1,300 @@
...
+ */
+'use strict';
+const Promise = require('bluebird');
+const db = require('../db/index');
+const IS = 'isDeleted';
+const AT = 'deletedAt';
+
+/**
+ * If a collector references a soft-deleted collector group, clear that
+ * collector group from the collector.
```
Comment (jgraff2) [inline migrations/20190520221710-no-paranoid.js]: Do we need to do this from the generator side as well?
Are there any other associations we need to do this for?

File: migrations/20190520221710-no-paranoid.js [hunk 5]
```diff
@@ -0,0 +1,300 @@
...
+    exec('Perspective', optsIs),
+    exec('Profile', optsIs),
+    exec('SSOConfig', optsAt),
+    exec('Token', optsIs),
+    exec('Collector', optsIs),
+    exec('GeneratorTemplate', optsIs),
+    exec('Lens', optsIs),
+    exec('Subject', optsIs),
+    exec('CollectorGroup', optsIs),
+  ]))
```
Comment (jgraff2) [inline migrations/20190520221710-no-paranoid.js]: Promise.all will run the promises in parallel, so it's possible they won't be executed in the right order.

File: db/model/user.js (hunk 2 above)
Comment (iamigo) [reply db/model/user.js]: Not likely to be a problem for us, but you're right. I'll keep paranoid on for profiles, too, for now.

File: tests/db/model/aspect/delete.js (hunk 3 above)
Comment (iamigo) [reply tests/db/model/aspect/delete.js]: Wow how'd that get in there?

File: migrations/20190520221710-no-paranoid.js (hunk 4 above)
Comment (iamigo) [reply migrations/20190520221710-no-paranoid.js]: These were the only ones where deletion was blocked for referential integrity in either sandbox or prod, I tried to think through other scenarios where similar things could happen but couldn't come up with any.

File: migrations/20190520221710-no-paranoid.js (hunk 5 above)
Comment (iamigo) [reply migrations/20190520221710-no-paranoid.js]: Yeah at one point I had a sequence of Promise.alls but then I wrapped the whole list in a trx so it all happens atomically.

**Comment by iamigo** [comment]
Thx for the great review, @jgraff2! I will make the changes on Monday.

**Pull request closed without merging**